### ISSUE TITLE
test_endpoints flaky with randomized hashes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5745]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5745|
|Type|defect|
|Created|2012-07-08 20:51:17Z|

One test depends on hash ordering and fails sporadically with ``python -R``:

```
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/home/antoine/twisted/twisted/internet/test/test_endpoints.py", line 1513, in test_ssl
    expectedCerts)
  File "/home/antoine/twisted/twisted/trial/unittest.py", line 270, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = [<Certificate Subject=fake-ca-2.example.com Issuer=fake-ca-2.example.com>,
 <Certificate Subject=fake-ca-1.example.com Issuer=fake-ca-1.example.com>]
b = [<Certificate Subject=fake-ca-1.example.com Issuer=fake-ca-1.example.com>,
 <Certificate Subject=fake-ca-2.example.com Issuer=fake-ca-2.example.com>]
```

Attaching patch.

Attachments:

* [randomhash_test_endpoints.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cfa/cfa3ba9f6fb47fa85a562d064d119b11810d61b8/cdcbe56b15719dc2ee618ba61690ccc352288728.patch) (810 bytes) - added by antoine on 2012-07-08 20:51:36Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5745 5745
type__defect defect
reporter__antoine antoine
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1341780677000000 1341780677000000
changetime__1341793440000000 1341793440000000
version__None None
owner__exarkun exarkun
cc__antoine
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @exarkun|
|-|-|

Fine, thanks.  For future readers, the dictionary gets involved in `_loadCAsFromDir`, which sticks certificates loaded from a CA directory (full of PEMs) into a dictionary to unique them.  I'll apply the patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

(In [34705]) Avoid relying on dict ordering when testing that the SSL endpoint loads CA certificates from the specified directory.

Author: antoine
Reviewer: exarkun
Fixes: [#5745](https://github.com/twisted/twisted/issues/5745)
